nixos/modules/misc/nixpkgs.nix

```diff
@@ -324,74 +324,152 @@ in
-        # Although, if nixpkgs.pkgs is set, we did forward it, but we did not construct it.
-          && !opt.pkgs.isDefined;
-    in [
-      (
-        let
-          nixosExpectedSystem =
-            if config.nixpkgs.crossSystem != null
-            then config.nixpkgs.crossSystem.system or (lib.systems.parse.doubleFromSystem (lib.systems.parse.mkSystemFromString config.nixpkgs.crossSystem.config))
-            else config.nixpkgs.localSystem.system or (lib.systems.parse.doubleFromSystem (lib.systems.parse.mkSystemFromString config.nixpkgs.localSystem.config));
-          nixosOption =
-            if config.nixpkgs.crossSystem != null
-            then "nixpkgs.crossSystem"
-            else "nixpkgs.localSystem";
-          pkgsSystem = finalPkgs.stdenv.targetPlatform.system;
-        in {
-          assertion = constructedByMe -> !hasPlatform -> nixosExpectedSystem == pkgsSystem;
-          message = "The NixOS nixpkgs.pkgs option was set to a Nixpkgs invocation that compiles to target system ${pkgsSystem} but NixOS was configured for system ${nixosExpectedSystem} via NixOS option ${nixosOption}. The NixOS system settings must match the Nixpkgs target system.";
-        }
-      )
-      {
-        assertion = constructedByMe -> hasPlatform -> legacyOptionsDefined == [];
-        message = ''
-          Your system configures nixpkgs with the platform parameter${optionalString hasBuildPlatform "s"}:
-          ${hostPlatformLine
-          }${buildPlatformLine
-          }
-          However, it also defines the legacy options:
-          ${concatMapStrings showOptionWithDefLocs legacyOptionsDefined}
-          For a future proof system configuration, we recommend to remove
-          the legacy definitions.
+        example = ''builtins.fetchTarball { name = "source"; sha256 = "${lib.fakeHash}"; url = "https://github.com/nixos/nixpkgs/archive/somecommit.tar.gz"; }'';
+
+        description = mdDoc ''
+          The path to the nixpkgs sources used to build the system as a string-with-context. This is
+          automatically set up to be the store path of the nixpkgs flake used to build the system if
+          using `nixpkgs.lib.nixosSystem`, and is otherwise null by default.
+
+          This can also be optionally set if the NixOS system is not built with a flake but still
+          uses pinned sources: set this to the store path for the nixpkgs sources used to build the
+          system, as may be obtained by `builtins.fetchTarball`, for example.
+
+          Note: the name of the store path must be "source" due to
+          <https://github.com/NixOS/nix/issues/7075>.
+        '';
+      };
+
+      setNixPath = mkOption {
+        type = types.bool;
+
+        default = cfg.flake.source != null;
+        defaultText = "config.nixpkgs.flake.source != null";
+
+        description = mdDoc ''
+          Whether to create `/etc/nix/inputs/nixpkgs` as a symlink to the store path of the nixpkgs
```

Comment:
`/etc/nix/inputs` does not seem to be the right place for this. I see the idea behind it, but then this should be used and established more broadly as a pattern. I can imagine a design where is some standard locations like XDG a user can easily see and access symlinks to inputs sources, assisting a way to keep those sources around via GC. But including such a concept here means it would be unique to NixOS, but less accessible on other systems.

Another possibility is to place it into the system derivation such that it will be in /run/current-system or /nix/var/nix/profiles/system. Etc is more stateful.

A possibility is to use `/nix/var/nix/profiles`?

Reply:
We can definitely do that. Putting a `/run/current-system/inputs/*` path in there feels pretty reasonable. But this is also somewhat revelatory of this change being better suited to a different structure, possibly .

It might be named, like, `system.inputs` or something? I am perusing `system/activation/top-level.nix` right now and it looks like to do this cleanly, we would have to do something somewhat similar to what bootspecs did; that is, poking at the system builder.

My feeling with this whole affair is that "nixos including a pinned nixpkgs version by nixos mechanisms" *is* a nixos only thing. It's not really obvious to me how you would do similar things on not-nixos; that is, setting `<nixpkgs>` to a nixpkgs path defined in a flake, with the tools we have, without some kind of weird script.

Reply:
The second thought I have with the last paragraph is that this is a symptom of a broader problem of the non-nixos experience with specifying things declaratively and pinning them generally being crap, and that `nix profile` is also not great. For now let's make NixOS better, but that issue is another story that should be improved.

Reply:
hmmm yeah, on non-nixos cases it does make sense to talk about the inputs of a profile, but not all profiles will necessarily have a specific, single nixpkgs input in the way that nixos configurations are built out of a specific nixpkgs (even if they also include packages from other ones).

like in particular, I don't use mutable state installation so I don't really know the details but I imagine it must be pretty common to end up with a mixed batch of packages from different versions, for people who do?

Reply:
These docs still reference the "/etc" location